spec/queries/keywords_query_spec.rb

```diff
@@ -3,44 +3,220 @@
+  context 'given no param' do
+    context 'given a user with no keywords' do
+      it 'returns an empty collection' do
+        result = described_class.new(Fabricate(:user).keywords).call
 
-      expect(result).to be_empty
+        expect(result[:keywords]).to be_empty
+      end
+    end
+
+    context 'given a user with 3 keywords' do
+      it 'returns exactly 3 keywords' do
+        user = Fabricate(:user)
+        Fabricate.times(3, :keyword, user: user)
+
+        result = described_class.new(user.keywords).call
+
+        expect(result[:keywords].length).to eq(3)
+      end
+
+      it 'orders the result by name ascending' do
+        user = Fabricate(:user)
+
+        %w[World awesome Pool].each { |name| Fabricate(:keyword, user: user, name: name) }
+
+        result = described_class.new(user.keywords).call
+
+        expect(result[:keywords].map(&:name)).to eq(%w[awesome Pool World])
+      end
+    end
+
+    context 'given many users with keywords' do
+      it 'returns only keywords from the initialized user' do
+        users = Fabricate.times(2, :user)
+        Fabricate.times(10, :keyword, user: users[0])
+        Fabricate.times(15, :keyword, user: users[1])
+
+        result = described_class.new(users[0].keywords).call
+
+        expect(result[:keywords].map(&:user_id)).to all(eq(users[0].id))
+      end
+
+      it 'only counts url matches from the initialized user' do
+        users = Fabricate.times(2, :user)
+        Fabricate.times(10, :keyword_parsed_with_links, user: users[0])
+        Fabricate.times(15, :keyword_parsed_with_links, user: users[1])
+
+        result = described_class.new(users[0].keywords, { url_pattern: '.*' }).call
+
+        expect(result[:url_match_count]).to eq(users[0].keywords.sum { |keyword| keyword.result_links.count })
+      end
+    end
+  end
+
+  context 'given a keyword_pattern' do
+    context 'given a name as keyword_pattern' do
+      it 'returns only keywords matching exactly that name' do
+        user = Fabricate(:user)
+
+        %w[World awesome Pool].each { |name| Fabricate(:keyword, user: user, name: name) }
+
+        result = described_class.new(user.keywords, { keyword_pattern: 'awesome' }).call
+
+        expect(result[:keywords].map(&:name)).to eq(%w[awesome])
+      end
+
+      it 'matches results in a case insensitive way' do
+        user = Fabricate(:user)
+
+        %w[World awesome Pool].each { |name| Fabricate(:keyword, user: user, name: name) }
+
+        result = described_class.new(user.keywords, { keyword_pattern: 'pooL' }).call
+
+        expect(result[:keywords].map(&:name)).to eq(%w[Pool])
+      end
+    end
+
+    context 'given an expression as keyword_pattern' do
+      it 'returns only keywords matching the given pattern' do
+        user = Fabricate(:user)
+
+        %w[World awesome Pool Zipline Zoo].each { |name| Fabricate(:keyword, user: user, name: name) }
+
+        result = described_class.new(user.keywords, { keyword_pattern: '.*o.*' }).call
```

Comment:
What if user submits an in valid `keyword_pattern` pattern? will it raise an exception? or it will return all keywords of the user as default?

Reply:
Yep, today it raise an exception! I will fix that with an alert message~ Nice catch @junan  :D

Reply:
## Update
After several trials, it is quite tricky to rescue the error:
- Cannot rescue the error inside the query object (as this one should NOT execute/load the keywords before pagination with pagy)
- Rescuing the error inside the controller is definitely not DRY and makes the actions quite heavy

I am currently implementing another solution: a simple REGEX to limitate the usage of the pattern, that will prevent having any invalid POSIX Expression send to our DB. This solution seems best because:
- We prevent the error instead of treating it
- Better for performance
- Better for code (everything can be done inside the query object then)

Reply:
## Update 2 
It is working fine, but I'm now having another question: 
> Should I put the filters params into a separate model and apply validation on that model? 
What do you think @ankitkalia1195 @junan ?

Thanks! :) 

PS: PR back to WIP so that I can add tests for the validation.

Reply:
I finally added another Model (`model/filter.rb`) to implement a clean validation of the filters, outside from the Query Object.